packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/textFeatures.ts

```diff
@@ -0,0 +1,146 @@
+                    const { ranges } = selection;
+                    const range = ranges[0];
+                    if (isWholeParagraphSelected(editor, range)) {
+                        setIndentation(editor, Indentation.Increase);
+                    } else {
+                        const tempRange = createRange(range.startContainer, range.startOffset);
+                        ranges.forEach(range => range.deleteContents());
+                        editor.select(tempRange);
+                        insertTab(editor, event);
+                    }
+                }
+            });
+
+            event.rawEvent.preventDefault();
+        }
+    },
+};
+
+/**
+ * @internal
+ */
+export const TextFeatures: Record<
+    keyof TextFeatureSettings,
+    BuildInEditFeature<PluginKeyboardEvent>
+> = {
+    indentWhenTabText: IndentWhenTabText,
+};
+
+function isWholeParagraphSelected(editor: IEditor, range: Range): boolean {
+    const endPosition: NodePosition = Position.getEnd(range);
+    const startPosition: NodePosition = Position.getStart(range);
+
+    const isAtStart: boolean = checkIfIsAtBlockLimit(editor, startPosition, true /* start */);
+    const isAtEnd: boolean = checkIfIsAtBlockLimit(editor, endPosition, false /* start */);
+
+    return isAtEnd && isAtStart;
+}
+
+/**
+ * Checks whether the position provided is at the start or at the end of the block element that contains the position.
+ * @param editor Editor Instance
+ * @param pos Position to check if is at the limits of the block
+ * @param start If true, checks from the beginning of the block, if false check from the end of block
+ * @returns boolean. true if is at limit, false otherwise
+ */
+function checkIfIsAtBlockLimit(editor: IEditor, pos: NodePosition, start: boolean): boolean {
+    let isAtLimit: boolean;
+    let blockElement = editor.getBlockElementAtNode(pos.node).getEndNode();
+    const block = (start ? blockElement.firstChild : blockElement.lastChild) || blockElement;
+
+    const tempChild = ignoreEmptyElement(block, start);
+
+    if (tempChild == pos.node && start ? pos.offset == 0 : pos.isAtEnd) {
+        isAtLimit = true;
+    }
+    return isAtLimit;
+}
+
+/**
+ * Ignore unselectable empty elements at start or end of the block
+ * @param block Block to check
+ * @param start if true, will check from the begin of the block, else will check the end of the block
+ * @returns Last node at the end or start
+ */
+function ignoreEmptyElement(block: Node | ChildNode, start: boolean): Node | ChildNode {
+    const getNextElement = (node: Node, block: Node | ChildNode) =>
+        start
+            ? node.nextSibling || block.nextSibling
+            : node.previousSibling || block.previousSibling;
+    const getElementToCheck = (node: Node) => (start ? node.firstChild : node.lastChild);
+
+    let tempChild = block;
+    while (getElementToCheck(tempChild)) {
+        tempChild = getElementToCheck(tempChild);
+    }
+
+    while (isEmptyElement(tempChild)) {
```

Comment:
Can this work correctly for image?

Reply:
I see, yes this did not work with a IMG at the end or at the start of the block element, but I think with the snippet you provided this is fixed.